docs/public_ips.md

```diff
@@ -0,0 +1,289 @@
+
+
+(These are just examples; it is valid to specify a hostPort with visiblity==loadbalancer, or to omit hostPort
+with visibility==public)
+
+In order to verify that the port can be assigned, the API server must check that the port is within a permitted range,
+and that it is not already assigned to another service.  This is similar to the allocation of internal cluster-IPs.
+The error codes & messages (in the case of port conflicts, out-of-range ports or port-exhaustion) will be analogous
+to the equivalent IP address errors.
+
+Advanced clouds (e.g. GCE) can support pure load-balancing without requiring a public port assignment, but
+for V1 visibility==loadbalancer incorporates visibility==public, so one will be assigned.
+
+## New flags
+
+We will add a flag `public_service_ports` to the apiserver, which will determine the port-range allowed for public services.
+
+The syntax is `--public_service_ports=30000-32767` to specify the range 30000-32767 _inclusive_.
+
+(Inside the code we will likely use a different representation)
+
+The default value is `--public_service_ports=30000-32767`.  This excludes most well-known ports, and certainly excludes
+the <1024 ports.  We should try to keep the minion using ports <1024.
+
+TODO: figure out which >1024 ports are needed on the minion
+
+A null range can be specified (e.g. `--public_service_ports=0-0`).  This will have the effect of preventing any public port allocations.
+
+In future, we could allow disjoint port ranges (`--public_service_ports=10000-20000,30000-32767`), but not for V1
+
+The range should not overlap the range of ports which kube-proxy uses internally for _cluster_ services;
+that range is set in /proc/sys/net/ipv4/ip_local_port_range.  We cannot know on the apiserver what the
+setting is on the kube-proxy machines (the minions), so we cannot prevent the user setting an overlapping range.
+The salt scripts could check this.  If an overlapping range is set, conflicts will randomly happen between
+cluster services and public services on some kube-proxy instances, and those kube-proxies will log errors and
+fail to serve that service.  This will likely result in degraded service availability.
+
+## API evolution
+
+The main changes:
+
+* we deprecate createExternalLoadBalancer, and replace it with visibility
+* we move publicIPs to the Status
+* we add a field hostPort to each service port
+ 
+Mappings:
+
+`visibility` is mappable:
+
+```
+new.visibility = old.createExternalLoadBalancer ? "loadbalancer" : "cluster"
+```
+
+```
+old.createExternalLoadBalancer = (new.visibility == 'loadbalancer')
+```
+
+`publicIPs` will be copied to the Spec if set on the Status, and from the Spec to the Status if not set on the Status.
+Once each cloud provider sets the Status, we will stop copying from the Spec to the Status.
+
+For v1 final:
+* we should remove publicIPs from the Spec, and require that this information is retrieved through the Status.
```

Comment:
> On May 6, 2015, at 8:45 PM, Brian Grant <notifications@github.com> wrote:
> 
> In docs/public_ips.md:
> 
> > +
> > +`visibility` is mappable:
> > +
> > +```
> > +new.visibility = old.createExternalLoadBalancer ? "loadbalancer" : "cluster"
> > +```
> > +
> > +```
> > +old.createExternalLoadBalancer = (new.visibility == 'loadbalancer')
> > +```
> > +
> > +`publicIPs` will be copied to the Spec if set on the Status, and from the Spec to the Status if not set on the Status.
> > +Once each cloud provider sets the Status, we will stop copying from the Spec to the Status.
> > +
> > +For v1 final:
> > +* we should remove publicIPs from the Spec, and require that this information is retrieved through the Status.
> This would mean that it wouldn't be possible for someone to set up a pool of VMs/machines with public IPs to run load balancers targeting the service endpoints, at least not without implementing a cloudprovider.
> 
> @smarterclayton Did you have any plans to use publicIPs for such a purpose?
> 
Yes.  We want to be able to run without a cloud provider and let administrators program the cluster (enforces the pattern that cloud providers operate from the outside in)
> —
> Reply to this email directly or view it on GitHub.
>